sdk/metric/internal/aggregate/sum.go

```diff
@@ -6,58 +6,92 @@ package aggregate // import "go.opentelemetry.io/otel/sdk/metric/internal/aggreg
 import (
 	"context"
 	"sync"
+	"sync/atomic"
 	"time"
 
 	"go.opentelemetry.io/otel/attribute"
 	"go.opentelemetry.io/otel/sdk/metric/metricdata"
 )
 
 type sumValue[N int64 | float64] struct {
-	n     N
+	n     atomicCounter[N]
 	res   FilteredExemplarReservoir[N]
 	attrs attribute.Set
 }
 
 // valueMap is the storage for sums.
 type valueMap[N int64 | float64] struct {
-	sync.Mutex
-	newRes func(attribute.Set) FilteredExemplarReservoir[N]
-	limit  limiter[sumValue[N]]
-	values map[attribute.Distinct]sumValue[N]
+	newRes   func(attribute.Set) FilteredExemplarReservoir[N]
+	aggLimit int
+
+	// cumulative sums do not reset values during collection, so in that case
+	// clearValuesOnCollection is false, hcwg is unused, and only values[0]
+	// and len[0] are used. All other aggregations reset on collection, so we
+	// use hcwg to swap between the hot and cold maps and len so measurements
+	// can continue without blocking on collection.
+	//
+	// see hotColdWaitGroup for how this works.
+	clearValuesOnCollection bool
+	hcwg                    hotColdWaitGroup
+	values                  [2]sync.Map
+	len                     [2]atomic.Int64
 }
 
-func newValueMap[N int64 | float64](limit int, r func(attribute.Set) FilteredExemplarReservoir[N]) *valueMap[N] {
+func newValueMap[N int64 | float64](
+	limit int,
+	r func(attribute.Set) FilteredExemplarReservoir[N],
+	clearValuesOnCollection bool,
+) *valueMap[N] {
 	return &valueMap[N]{
-		newRes: r,
-		limit:  newLimiter[sumValue[N]](limit),
-		values: make(map[attribute.Distinct]sumValue[N]),
+		newRes:                  r,
+		aggLimit:                limit,
+		clearValuesOnCollection: clearValuesOnCollection,
 	}
 }
 
 func (s *valueMap[N]) measure(ctx context.Context, value N, fltrAttr attribute.Set, droppedAttr []attribute.KeyValue) {
-	s.Lock()
-	defer s.Unlock()
-
-	v, ok := s.values[fltrAttr.Equivalent()]
+	hotIdx := uint64(0)
+	if s.clearValuesOnCollection {
+		hotIdx = s.hcwg.start()
+		defer s.hcwg.done(hotIdx)
+	}
+	v, ok := s.values[hotIdx].Load(fltrAttr.Equivalent())
 	if !ok {
-		fltrAttr = s.limit.Attributes(fltrAttr, s.values)
-		v = s.values[fltrAttr.Equivalent()]
-		v.res = s.newRes(fltrAttr)
-		v.attrs = fltrAttr
+		// It is possible to exceed the attribute limit if it races with other
+		// new attribute sets. This is an accepted tradeoff to avoid locking
+		// for writes.
+		if s.aggLimit > 0 && s.len[hotIdx].Load() >= int64(s.aggLimit-1) {
+			fltrAttr = overflowSet
+		}
```

Comment:
A pre-increment approach with rollback may address this.

```suggestion
		// Pre-increment, reserve a slot first.
		newLen := s.len[hotIdx].Add(1)
		if s.aggLimit > 0 && newLen >= int64(s.aggLimit) {
			// We exceeded the limit, decrement back and use overflow
			s.len[hotIdx].Add(-1)
			fltrAttr = overflowSet
		}
```

Reply:
nice.  lgtm

Reply:
@dashpole thoughts: https://github.com/dashpole/opentelemetry-go/pull/2?